x-pack/filebeat/input/awss3/states.go

```diff
@@ -105,12 +107,13 @@ func (s *states) Delete(id string) {
 func (s *states) IsListingFullyStored(listingID string) bool {
 	info, _ := s.listingInfo.Load(listingID)
 	listingInfo := info.(*listingInfo)
-	if listingInfo.finalCheck {
+	if listingInfo.finalCheck.Load() {
 		return false
 	}
 
-	listingInfo.finalCheck = (listingInfo.storedObjects + listingInfo.errorObjects) == listingInfo.totObjects
-	return listingInfo.finalCheck
+	finalCheck := (listingInfo.storedObjects.Load() + listingInfo.errorObjects.Load()) == listingInfo.totObjects
```

Comment:
If `storedObjects` and `errorObjects` are int types, then they can be sanity checked for ≥ 0 prior to the summation. Is it ever possible for the sum of stored and errored to be greater than total? Testing for exact equality makes me queasy in situations like this.

I'm a little concerned about the possibility of a now silent data race; the ensemble of `storedObjects` and `errorObjects` are not treated atomically, so it is possible for the loaded value of one of them to be come from one execution state and the other from another. This would never be detected by the race detector since now they are guarded by write barriers. I can't see a situation where this would have a bad outcome, but I think it is probably better to either lock the whole thing of make the access properly atomic using https://pkg.go.dev/sync/atomic#Value (the complexity of that probably argues for locks).

Reply:
I will then revert to the first commit, where everything was treated atomically

Reply:
>Is it ever possible for the sum of stored and errored to be greater than total?

not unless of a bug in the code

Reply:
> not unless of a bug in the code

That's worth catching and logging then, no? 😄